cunumeric/deferred.py

```diff
@@ -441,27 +688,63 @@ def set_item(self, key, rhs):
         assert self.dtype == rhs.dtype
         # Check to see if this is advanced indexing or not
         if is_advanced_indexing(key):
+            view_copy = False
             # Create the indexing array
-            index_array = self._create_indexing_array(key)
-            if index_array.shape != rhs.shape:
-                raise ValueError(
-                    "Advanced indexing array does not match source shape"
-                )
-            if self.ndim != index_array.ndim:
-                raise NotImplementedError(
-                    "need support for indirect partitioning"
+            copy_needed, store, index_array = self._create_indexing_array(
+                key, True
+            )
+            if copy_needed:
+                if self.base.transform.bottom:
+                    lhs = self
+                else:
+                    # if  store is transformed we need to to return a copy of
+                    # the store since Copy operation can't be done on
+                    # the store with transformation
+                    store_to_copy = DeferredArray(
+                        self.runtime,
+                        base=store,
+                        dtype=self.dtype,
+                    )
+                    store_copy = self.runtime.create_empty_thunk(
+                        store_to_copy.shape,
+                        self.dtype,
+                        inputs=[store_to_copy],
+                    )
+                    store_copy.copy(store_to_copy, deep=True)
+
+                    lhs = store_copy
+                    view_copy = True
+            else:
+                lhs = self
+                view_copy = False
+
+            if rhs.ndim == 0:
```

Comment:
Did the `_broadcast` case below not work for `rhs.ndim == 0`?

Reply:
`_ broadcast` doesn't seem to convert scalar to `store`. If I try using it, I get an error:
```
: python exception occurred within task:
Traceback (most recent call last):
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legion_top.py", line 410, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legion_top.py", line 234, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "tests/index_routines.py", line 582, in <module>
    test()
  File "tests/index_routines.py", line 576, in test
    advanced_indexing()
  File "tests/index_routines.py", line 194, in advanced_indexing
    y_num[index_num] = 3
  File "/home/idemeshko/development/cunumeric_ai/cunumeric/array.py", line 139, in wrapper
    return func(*args, **kwargs)
  File "/home/idemeshko/development/cunumeric_ai/cunumeric/array.py", line 1436, in __setitem__
    self._thunk.set_item(key, value._thunk)
  File "/home/idemeshko/development/cunumeric_ai/cunumeric/deferred.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/home/idemeshko/development/cunumeric_ai/cunumeric/deferred.py", line 736, in set_item
    copy.execute()
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/operation.py", line 183, in execute
    self._context.runtime.submit(self)
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/runtime.py", line 944, in submit
    self.flush_scheduling_window()
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/runtime.py", line 939, in flush_scheduling_window
    self._schedule(ops)
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/runtime.py", line 932, in _schedule
    op.launch(strategy)
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/operation.py", line 686, in launch
    launcher.add_input(store, req, tag=tag)
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/launcher.py", line 982, in add_input
    self.add_store(store, proj, Permission.READ, tag, flags)
  File "/home/idemeshko/install/legate/lib/python3.8/site-packages/legate/core/launcher.py", line 969, in add_store
    assert store.kind is not Future
AssertionError
legion_python: /home/idemeshko/development/legate.core/legion/runtime/realm/python/python_module.cc:992: virtual void Realm::LocalPythonProcessor::execute_task(Realm::Processor::TaskFuncID, const Realm::ByteArrayRef&): Assertion `0' failed.
Signal 6 received by node 0, process 1129888 (thread 7f9a64fd1000) - obtaining backtrace
```

Reply:
Can you make a small reproducer for this and post as a bug on legate.core? I think a scalar `Store` is still a `Store`, and `_broadcast` isn't doing anything outside of the ordinary for `Store`s. And please add a TODO if you can, to update this once the bug is fixed.

Reply:
@manopapad : will do.

Reply:
Was this fixed?

Reply:
Yes, the issue was that res wan't converted to the deferred array automatically. It is fixed now